Extracting SubjectKeyId and AuthorityKeyId in case of x509 V3 extensions v.2

mprse opened the pull request:
## Description

This PR is continuation of PR https://github.com/Mbed-TLS/mbedtls/pull/4161:
- rebased on `development`
- fixed build errors
- removed redundant code to generate cert from makefile


Extracting SubjectKeyId and AuthorityKeyId in case of x509 V3 extensions. Updating mbedtls_x509_crt_free function to also free the new dynamic elements (issuer field of AuthorityKeyId). Resolves https://github.com/Mbed-TLS/mbedtls/issues/4666.

A few tests are also added which test the feature with a correct certificate and multiple ones with erroneous ASN1 tags.

Needs: https://github.com/Mbed-TLS/mbedtls/pull/6882 https://github.com/Mbed-TLS/mbedtls/pull/6900


## Gatekeeper checklist

- [x] **changelog** provided
- [x] **backport** not required - it's a new feature
- [x] **tests** provided


## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

AndrzejKurek commented on tests/data_files/Makefile:
```diff
@@ -318,6 +318,15 @@ rsa_pkcs8_2048_public.der: rsa_pkcs8_2048_public.pem
+authorityKeyId_subjectKeyId.crt:
+	$(OPENSSL) req -x509 -nodes -days 7300 -newkey rsa:2048 -keyout authorityKeyId_subjectKeyId.crt -out authorityKeyId_subjectKeyId.crt -config authorityKeyId_subjectKeyId.conf -extensions 'v3_req'
+# The listed certificates are the copies of authorityKeyId_subjectKeyId.crt with error injections
+# authorityKeyId_subjectKeyId_wrong_SubjectKeyId.crt              The TAG marking the beginning of SubjectKeyId is set to 0x00
```
It would be great to have makefile rules to create each of these files. 
See https://github.com/Mbed-TLS/mbedtls/pull/6900/files#diff-33f05cba6c4be0b2bf38b859d772b1a0527c9d80ad7347aada68eee172bb584aR297 for reference.

mprse replied:
Yes, I agree that this would be better solution, but this would require generating new certs and malformed versions.
I suggest to leave it for now and create issue for this.

AndrzejKurek commented on include/mbedtls/x509_crt.h:
```diff
@@ -598,7 +601,7 @@ int mbedtls_x509_crt_parse_path(mbedtls_x509_crt *chain, const char *path);
  *                 of the subject alternative name encoded in \p san_raw.
  *
- * \note           Only "dnsName" and "otherName" of type hardware_module_name
+ * \note           Only "dnsName" and "otherName" and "rfc822Name" of type hardware_module_name
```
This description is wrong. "hardware_module_name" is a subtype of "otherName", and "rfc822Name" is on the same level as "otherName". Also, while we're at it, I believe that the next line should mention RFC 5280, not 4180.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -634,8 +634,36 @@ static int x509_get_ext_key_usage(unsigned char **p,
 }
 
 /*
- * SubjectAltName ::= GeneralNames
```
Missing line

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -634,8 +634,36 @@ static int x509_get_ext_key_usage(unsigned char **p,
+    }
+
+    if (*p != end) {
+        return MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
```
Shouldn't this now be MBEDTLS_ERROR_ADD?

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -691,10 +719,14 @@ static int x509_get_subject_alt_name(unsigned char **p,
-                                     MBEDTLS_ERR_ASN1_UNEXPECTED_TAG);
+            if ((tag & (MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE)) !=
+                (MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE)) {
+                return MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
```
Same problem: `MBEDTLS_ERROR_ADD` dropped

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +777,83 @@ static int x509_get_subject_alt_name(unsigned char **p,
+            /* Getting directoryName using the required specific class tag [4] */
+            if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                            MBEDTLS_ASN1_CONTEXT_SPECIFIC |
+                                            MBEDTLS_ASN1_CONSTRUCTED | 4)) != 0) {
```
use `MBEDTLS_X509_SAN_DIRECTORY_NAME` instead of hardcoded 4

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +777,83 @@ static int x509_get_subject_alt_name(unsigned char **p,
+ *        authorityCertIssuer [1] GeneralNames OPTIONAL,
+ *        authorityCertSerialNumber [2] CertificateSerialNumber OPTIONAL }
+ *
+ *    KeyIdentifier ::= OCTET STRING
```
Missing information, check, and also a test for:
```
 -- authorityCertIssuer and authorityCertSerialNumber MUST both
    -- be present or both be absent
```

mprse replied:
Adding extra code to the parser to verify that both `authorityCertIssuer`, `authorityCertSerialNumber` are present or not doesn't matter too much. Also extra code needs to be added. For parser we should ensure that it won't crash if there is a case where for example `authorityCertIssuer` is present and `authorityCertSerialNumber` is not. From this perspective I think the current code is correct.

Function that generates or  verifies certificates should detect this case and return error.

AndrzejKurek replied:
You're right!

DemiMarie replied:
It’s possible to implicitly enforce this check: parse `authorityCertSerialNumber` as a mandatory field if `authorityCertIssuer` is present, and do not parse it at all if `authorityCertIssuer` is absent.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +777,83 @@ static int x509_get_subject_alt_name(unsigned char **p,
+    size_t len = 0u;
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                    MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE | 0)) != 0) {
```
`MBEDTLS_OID_X509_EXT_AUTHORITY_KEY_IDENTIFIER` instead of 0?

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -1839,6 +1965,31 @@ int mbedtls_x509_parse_subject_alt_name(const mbedtls_x509_buf *san_buf,
+         */
+        case (MBEDTLS_ASN1_SEQUENCE | MBEDTLS_X509_SAN_RFC822_NAME):
+        {
+            mbedtls_x509_name rfc822Name;
```
I think that it would be best to split this PR into two. This is for parsing an additional subject alt name type, even though the PR name (or description) mentions it. Also, there are no tests for this code.

AndrzejKurek reviewed (changes requested):
A blocker for me right now is the fact that this PR adds parsing of RFC822 name with no tests - please extract it to a separate PR (I can handle that if we agree on that and you'll remove the code from here). Additionally - please use files instead of raw data in .data file and ensure that coverage is checked as in other work we're pushing.

mprse commented:
Addressed review comments and created additional tasks as agreed in comments:
- [improve testing](https://github.com/Mbed-TLS/mbedtls/issues/6985)
- [add parsing of x509 RFC822 name](https://github.com/Mbed-TLS/mbedtls/issues/6986)

Test coverage for the new code is fine. Ready for review.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +778,84 @@ static int x509_get_subject_alt_name(unsigned char **p,
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t len = 0u;
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
```
If this check gets a length that's outside of bounds, it's silently ignored.
Please add a test for that and also check, if 
```
    if (*p + len != end) {
        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS,
                                 MBEDTLS_ERR_ASN1_LENGTH_MISMATCH);
    }
```

mprse replied:
Fixed.

Also adapted returned errors to indicate extension parsing failure (`MBEDTLS_ERR_X509_INVALID_EXTENSIONS`).

BhanuPrakash-P replied:
We have an application which uses tiny-curl, mbedtls for SSL connections.
On our ESP-IDF v5.0.1, SSL connections were fine and able to communicate with Azure IoT.
Recently, we started upgrading ESP-IDF to v5.1.2 with mbedtls v3.5.0, we observed an error for "AuthorityKeyIdentifier" as follows.

Curl Debug Callback()[0]: * Error reading private key (nil) - mbedTLS: (-0x2566) X509 - The extension tag or value is invalid : ASN1 - Actual length differs from expected length

What are the possible cases when length can go outside of bounds?

gilles-peskine-arm replied:
@BhanuPrakash-P In principle this error indicates that you have some bad data inside an X.509 structure. But X.509 is complicated and I can't rule out a bug or limitation in Mbed TLS. If you want to have any chance of getting help, please post an issue with an actual X.509 certificate and a small amount of code (or maybe one of the [sample programs](https://github.com/Mbed-TLS/mbedtls/tree/mbedtls-3.6.0/programs/x509)) that reproduce the problem.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -691,10 +720,14 @@ static int x509_get_subject_alt_name(unsigned char **p,
             MBEDTLS_ASN1_CONTEXT_SPECIFIC) {
-            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS,
-                                     MBEDTLS_ERR_ASN1_UNEXPECTED_TAG);
+            if ((tag & (MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE)) !=
```
Unfortunately, this code shouldn't be here. I'll post bigger explanations below.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +778,89 @@ static int x509_get_subject_alt_name(unsigned char **p,
+
+    if (*p < end) {
+        /* Getting authorityCertIssuer using the required specific class tag [1] */
+        if ((ret = mbedtls_asn1_get_tag(p, end, &len,
```
Unfortunately this code is wrong. It should not be checking for `MBEDTLS_X509_SAN_RFC822_NAME` here, but for the ` authorityCertIssuer       [1]` context-specific choice (which also happens to be 1).

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +778,89 @@ static int x509_get_subject_alt_name(unsigned char **p,
+            /* authorityCertIssuer is an OPTIONAL field */
+        } else {
+            /* Getting directoryName using the required specific class tag [4] */
+            if ((ret = mbedtls_asn1_get_tag(p, end, &len,
```
This is also wrong. we shouldn't be looking so deep, this check is already determining the subtype of a generalName.

DemiMarie replied:
What do you mean?

mprse replied:
As far as I remember the original code here tried to parse `GeneralName` manually and incorrectly. We have introduced `mbedtls_x509_get_subject_alt_name_ext` for this purpose. 

This part of code is valid. The original code below this was determining the subtype of a `generalName`.

AndrzejKurek replied:
Yes, this comment is long resolved and does not apply to the current state of the PR.

AndrzejKurek commented on library/x509_crt.c:
```diff
@@ -745,6 +778,89 @@ static int x509_get_subject_alt_name(unsigned char **p,
+                return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+            } else {
+                /* "end" also includes the CertSerialNumber field so "len" shall be used */
+                ret = x509_get_general_names(p,
```
This passes a pointer to a structure too deep inside the cert. There's also no easy fix, since for the authorityKeyID there's no external sequence number, so the functions have to be reworked. This funcion was also failing internally, but it's hidden and unchecked.
![image](https://user-images.githubusercontent.com/29195988/216310880-428cf57e-8d43-4360-bac3-4831d2c9e9aa.png)

mprse commented:
Rebased after https://github.com/Mbed-TLS/mbedtls/pull/6882. Still https://github.com/Mbed-TLS/mbedtls/pull/6900 is required (authorityCertIssuer is of GenerlNames type and one of GenerlNames subtypes is DirectoryName which is used in provided tests).

mprse commented:
Rebased to resolve conflicts (`include/mbedtls/x509_crt.h`) and fixed test dependencies in https://github.com/Mbed-TLS/mbedtls/pull/6866/commits/0a09bfc128c50eaaca07e9da59c38086b83f427d.

AndrzejKurek commented:
Adding `needs-work` due to CI failures.

mprse commented:
The following patch is required to fix issue with pksc7 test on CI: https://github.com/Mbed-TLS/mbedtls/pull/7394

mprse commented:
This one is now ready for review.

mprse commented:
While reviewing please note that test are improved n follow-up PR: https://github.com/Mbed-TLS/mbedtls/pull/7411

AndrzejKurek reviewed (changes requested):
Some comments left, but I have one bigger remark to wonder about here:
We're requiring the authorityCertSerialNumber field.
Please add a test with an empty authorityKeyIdentifier extension - it should work, right? All of the fields are optional.

mprse commented:
> Please add a test with an empty authorityKeyIdentifier extension - it should work, right? All of the fields are optional.

Good point, but I would like to do this in the follow-up PR where tests are reworked.

AndrzejKurek commented:
> > Please add a test with an empty authorityKeyIdentifier extension - it should work, right? All of the fields are optional.
> 
> Good point, but I would like to do this in the follow-up PR where tests are reworked.

OK, no problem, then a blocker for me is just the fact that the behaviour is not correct now - `authorityCertSerialNumber` has to be present, while, according to the spec, it could be omitted.

mprse commented on library/x509_crt.c:
```diff
@@ -586,6 +586,113 @@ static int x509_get_ext_key_usage(unsigned char **p,
+                                        2)) !=
+            0) {
+            /* authorityCertSerialNumber is an OPTIONAL field */
+            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
```
Good point. I think this was added by mistake.

mprse replied:
I tried to investigate the test failures and I think now that the original code was correct.
All tree fields are optional but if we are in `x509_get_authority_key_id`, then we have authority key identifier extension available.
Now we are trying to read the data inside the extension and if data does not match to keyIdentifier, authorityCertIssuer nor authorityCertSerialNumber then we report error.

AndrzejKurek reviewed (dismissed):
Looks good if we follow up with test improvements.

daverodgman commented:
@d3zd3z would you be able to take a look on review please?

DemiMarie commented on tests/suites/test_suite_x509parse.function:
```diff
@@ -1472,3 +1472,83 @@ exit:
+
+    mbedtls_x509_crt_init(&crt);
+
+    TEST_ASSERT(mbedtls_x509_crt_parse_der(&crt, buf->x, buf->len) == ref_ret);
```
Use `TEST_EQUAL` when asserting equality.  `TEST_EQUAL` prints both the expected and actual values, which makes debugging failed tests much easier.  I found this out the hard way when debugging #6659.

```suggestion
    TEST_EQUAL(mbedtls_x509_crt_parse_der(&crt, buf->x, buf->len), ref_ret);
```

(and same for all other uses of `TEST_ASSERT` on an equality)

mprse replied:
I agree, but I will fix this in follow-up PR with test improvements(https://github.com/Mbed-TLS/mbedtls/pull/7411).

DemiMarie commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -3165,3 +3165,35 @@ x509_accessor_ext_types:MBEDTLS_X509_EXT_KEY_USAGE:MBEDTLS_X509_EXT_KEY_USAGE
 depends_on:MBEDTLS_X509_CRT_PARSE_C
 x509_accessor_ext_types:MBEDTLS_X509_EXT_KEY_USAGE:MBEDTLS_X509_EXT_SUBJECT_ALT_NAME
+
+X509 CRT parse Subject Key Id - Correct Subject Key ID
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_subjectkeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341 ...
+
+X509 CRT parse Subject Key Id - Wrong OCTET_STRING tag
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_subjectkeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341 ...
+
+X509 CRT parse Authority Key Id - Correct Authority Key ID
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
+
+X509 CRT parse Authority Key Id - Wrong Length
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
+
+X509 CRT parse Authority Key Id - Wrong Sequence tag
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
+
+X509 CRT parse Authority Key Id - Wrong KeyId Tag
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
+
+X509 CRT parse Authority Key Id - Wrong Issuer Tag 1
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
+
+X509 CRT parse Authority Key Id - Wrong Issuer Tag 2
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"308203873082026fa003020102020100300d06092a864886f70d0101050500303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c2054657374204341301e170d3131303231323134343430305a170d3231303231323134343430305a303b310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c3119301706035504031310506f6c617253534c20546573742043 ...
```
How were the test certificates generated?  I want to make sure that they don’t have any other problems, as future changes (such as #6659) may make the ASN.1 parsing stricter.

AndrzejKurek replied:
This will be fixed [here](https://github.com/Mbed-TLS/mbedtls/pull/7411).

mprse commented on library/x509_crt.c:
```diff
@@ -1545,6 +1668,21 @@ int mbedtls_x509_crt_parse_path(mbedtls_x509_crt *chain, const char *path)
+
+#define KEY_USAGE(code, name)    \
+    if (key_usage & (code))    \
+    PRINT_ITEM(name);
```
Fixed.
BTW the line continuation backslashes are messy because our code-style auto-formatter.

d3zd3z reviewed (commented):
Just two little things, one a question, and the other wondering why we're using an else block in this instance.

mprse commented on library/x509_crt.c:
```diff
@@ -586,6 +586,112 @@ static int x509_get_ext_key_usage(unsigned char **p,
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                    MBEDTLS_ASN1_OCTET_STRING)) != 0) {
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+    } else {
```
Yes, it looks like this code can be optimized and `else` can be removed, but I think it is here for consistency. The same pattern is used in `x509_get_authority_key_id`. Sometimes no error is returned (if the field is optional) and then we have comment provided in the block for error.

mprse commented on programs/ssl/ssl_client2.c:
```diff
@@ -41,7 +41,7 @@ int main(void)
 /* Size of memory to be allocated for the heap, when using the library's memory
  * management and MBEDTLS_MEMORY_BUFFER_ALLOC_C is enabled. */
-#define MEMORY_HEAP_SIZE      120000
+#define MEMORY_HEAP_SIZE      180000
```
Good point I forgot to handle this part.

The discussion from the original PR:
https://github.com/Mbed-TLS/mbedtls/pull/4161/files#r608810763

To be honest I don't know what to do with this issue. As @gilles-peskine-arm suggested, the certificates for tests are provided in DER format. But I don't understand how this influences the `ssl_client2` program

We also don't know what should fail if the heap is too small. I suspect that some of `ssl-opt` tests should fail. We can try to decrease this and check what CI says.

gilles-peskine-arm replied:
`ssl_client2` and `ssl_server2` are test programs. We don't promise any kind of backward compatibility for well-defined interfaces such as command line arguments, let alone more fuzzy considerations such as heap usage (which is platform-dependent).

There is some value in being able to run these programs from embedded systems, but a given system might only need a given configuration which doesn't need as much memory (e.g. no RSA → smaller certificates), so this single number isn't what you'd want anyway. Note that `MEMORY_HEAP_SIZE` is only used when the malloc implementation from mbedtls is enabled (`MBEDTLS_MEMORY_BUFFER_ALLOC_C`), anyway.

mprse commented:
CI seems fine. @d3zd3z can you please review?

d3zd3z reviewed (approved):
Looks good. The else thing was just an aesthetic thing, so I'm not too concerned about it.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,112 @@ static int x509_get_ext_key_usage(unsigned char **p,
+        if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                        MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED |
+                                        1)) != 0) {
+            /* authorityCertIssuer is an OPTIONAL field */
```
This is an error condition.  If `*p < end` then both `authorityCertIssuer` and `authorityCertSerialNumber` must be present.

mprse replied:
I'm not sure if we want to do this for parsing. For parsing we should get as much as possible. Simply one field will be unavailable if not present, but parsing will succeed.

mprse replied:
Fixed

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,112 @@ static int x509_get_ext_key_usage(unsigned char **p,
+            /* authorityCertIssuer is an OPTIONAL field */
+        } else {
+            /* "end" also includes the CertSerialNumber field so "len" shall be used */
+            ret = mbedtls_x509_get_subject_alt_name_ext(p,
+                                                        (*p+len),
+                                                        &authority_key_id->authorityCertIssuer);
```
Missing return value check.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -1545,6 +1668,25 @@ int mbedtls_x509_crt_parse_path(mbedtls_x509_crt *chain, const char *path)
+#define KEY_USAGE(code, name)      \
+    do {                           \
+        if (key_usage & (code))    \
+        PRINT_ITEM(name);          \
```
```suggestion
            PRINT_ITEM(name);      \
```

mprse commented:
@DemiMarie Adapted code to fulfil requirement: `authorityCertIssuer and authorityCertSerialNumber MUST bothbe present or both be absent` and addressed remaining comments. Please approve the PR if you are fine with these changes.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+
+            /* Getting authorityCertSerialNumber using the required specific class tag [2] */
+            if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                            MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_INTEGER |
+                                            2)) != 0) {
```
```suggestion
                                            MBEDTLS_ASN1_CONTEXT_SPECIFIC | 2)) != 0) {
```

`MBEDTLS_ASN1_INTEGER` is 2 which is why this passed testing.

mprse commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+            } else {
+                authority_key_id->authorityCertSerialNumber.len = len;
+                authority_key_id->authorityCertSerialNumber.p = *p;
+                authority_key_id->authorityCertSerialNumber.tag = MBEDTLS_ASN1_OCTET_STRING;
```
From rfc5280:
`CertificateSerialNumber  ::=  INTEGER`

So should be `MBEDTLS_ASN1_INTEGER`.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                    MBEDTLS_ASN1_CONTEXT_SPECIFIC)) != 0) {
+        /* KeyIdentifier is an OPTIONAL field */
```
If `ret` is anything other than 0 or `MBEDTLS_ERR_ASN1_UNEXPECTED_TAG`, the value of `*p` and `len` is undefined.  Therefore, an explicit check for `MBEDTLS_ERR_ASN1_UNEXPECTED_TAG` is needed.

DemiMarie replied:
I don’t see any reason why `MBEDTLS_ERR_ASN1_OUT_OF_DATA` is not possible.  It will happen if the ASN.1 claims to have more bytes that it are actually present.  In this case, `p` will have been bumped even though `mbedtls_asn1_get_tag()` failed.

Anything return value other than 0 or `MBEDTLS_ERR_ASN1_UNEXPECTED_TAG` needs to cause an error to be returned.

mprse replied:
Fixed as suggested.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+            /* authorityCertIssuer and authorityCertSerialNumber MUST both
+               be present or both be absent. At this point we expect to have both. */
+            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+        } else {
+            /* "end" also includes the CertSerialNumber field so "len" shall be used */
+            ret = mbedtls_x509_get_subject_alt_name_ext(p,
+                                                        (*p+len),
+                                                        &authority_key_id->authorityCertIssuer);
+            if (ret != 0) {
+                return ret;
+            }
+
+            /* Getting authorityCertSerialNumber using the required specific class tag [2] */
+            if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                            MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_INTEGER |
+                                            2)) != 0) {
+                return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+            } else {
+                authority_key_id->authorityCertSerialNumber.len = len;
+                authority_key_id->authorityCertSerialNumber.p = *p;
+                authority_key_id->authorityCertSerialNumber.tag = MBEDTLS_ASN1_OCTET_STRING;
+                *p += len;
+            }
+        }
```
This `else` adds no real value here.

mprse replied:
Removed redundant `else` blocks.

mprse commented:
@DemiMarie Please confirm if review comments are addressed by approving this PR.

DemiMarie reviewed (approved):
The library code looks good.  I did not review the test data and did not throroughly review the test code.

AndrzejKurek reviewed (approved):
Looks good. Looking forward to test improvements that will include all the optional field configurations.

mprse commented:
This one is ready for merge.

gilles-peskine-arm commented:
> I did not review the test data and did not throroughly review the test code.

We do need two reviewers on the whole pull request, including the tests. I only see a full approval from @AndrzejKurek and a partial approval from @DemiMarie, so someone needs to do a second review of the tests.

We don't need to pay as much attention to the test code as to the library code, but we need to pay _some_ attention. And for the test data, several things are important in review:

* Verify that the test descriptions are accurate (e.g. to avoid copy-paste mistakes).
* Verify that the test data is correct (and not e.g. written to pass the tests without being confident that they're correct).
* Verify that we understand how the test data was created (e.g. so that we can create similar tests if we discover edge cases that need testing in the future).
* Verify that the test coverage is satisfactory.

gilles-peskine-arm commented:
Ah, @mprse has pointed me to [@d3zd3z's prior approval](https://github.com/Mbed-TLS/mbedtls/pull/6866#pullrequestreview-1399343629) of 94cf710edcf406c35d1c0c66560c9fe293667132 so all that remains is reviewing the last three commits since then from the perspective of tests.

I've just done so (not really understanding the feature or the library code) and I have one concern: the commit "Fix parsing of KeyIdentifier (tag error case)" seems to be fixing a bug that should have at least one non-regression test, but that commit doesn't add a test case. Is it covered by a test case (and so failed an earlier CI run)? Or does it need a test case?

gilles-peskine-arm commented:
After discussing with Przemysław: I'm requesting the addition of a non-regression test for "Fix parsing of KeyIdentifier (tag error case)". It's ok if it's generated from code that won't be included or reviewed until https://github.com/Mbed-TLS/mbedtls/pull/7411 goes through. But in case we do further maintenance on the new feature, or we do a release earlier than planned, I want to have the test case in with the feature.

mprse commented:
Force pushed the last commit to include test case. Also adapted returned error code.
@d3zd3z @AndrzejKurek @DemiMarie Can you please take a look again?

AndrzejKurek reviewed (dismissed):
Confirmed locally that the new error is returned (and the line is hit by a breakpoint) in the added test. If there will be a Makefile recipe for this in the next PR - I'm happy.

gilles-peskine-arm commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -3190,6 +3190,10 @@ X509 CRT parse Authority Key Id - Wrong KeyId Tag
 
+X509 CRT parse Authority Key Id - Wrong KeyId Tag Length
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_RSA_C
+x509_crt_parse_authoritykeyid:"30820214308201B9A00302010202143960EFDE5674DE1F7B761699CF8E5C024E209452300A06082A8648CE3D040302303B310B3009060355040613024E4C3111300F060355040B0C08506F6C617253534C3119301706035504030C10506F6C617253534C2054657374204341301E170D3233303430363037313834315A170D3433303430313037313834315A303B310B3009060355040613024E4C3111300F060355040B0C08506F6C617253534C3119301706035504030C ...
```
The test case is failing in some configurations. Looks like a missing dependency.

gilles-peskine-arm replied:
Apart from the dependencies, I'm happy that this is a non-regression test for the change in "Fix parsing of KeyIdentifier (tag length error case)". I have not verified that the change is correct, just that the library change is consistent with the test change.

mprse replied:
Ups.. I created certificate signed using EC key and this is the reason of failure. Other certificates are using RSA keys, so I force pushed fixed RSA signed cert.

gilles-peskine-arm commented on tests/suites/test_suite_x509parse.function:
```diff
@@ -1472,3 +1472,83 @@ exit:
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_X509_CRT_PARSE_C */
+void x509_crt_parse_subjectkeyid(data_t *buf, unsigned int subjectKeyIdLength, int ref_ret)
```
Sorry, you can't use `unsigned int` as the type of a test function argument. Until yesterday, the test framework was sloppy and accepted `unsigned int` but treated it as `int` under the hood. Since #6716 you can use larger _signed_ integer types (up to `intmax_t`), but unsigned types are rejected. A follow-up will make unsigned types work, but it's not ready yet.

mprse replied:
Modified test function parameters to `int`.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+                                    MBEDTLS_ASN1_OCTET_STRING)) != 0) {
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+    } else {
+        subject_key_id->len = len;
+        subject_key_id->tag = MBEDTLS_ASN1_OCTET_STRING;
+        subject_key_id->p = *p;
+        *p += len;
+    }
```
```suggestion
    }

    subject_key_id->len = len;
    subject_key_id->tag = MBEDTLS_ASN1_OCTET_STRING;
    subject_key_id->p = *p;
    *p += len;
```

Even when the field is optional, the return value must still be checked.

DemiMarie commented on library/x509_crt.c:
```diff
@@ -586,6 +586,114 @@ static int x509_get_ext_key_usage(unsigned char **p,
+                               MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+
+    /* KeyIdentifier is an OPTIONAL field */
+    if (ret != 0 && ret != MBEDTLS_ERR_ASN1_UNEXPECTED_TAG) {
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
+    } else if (ret == 0) {
+        authority_key_id->keyIdentifier.len = len;
+        authority_key_id->keyIdentifier.p = *p;
+        /* Setting tag of the keyIdentfier intentionally to 0x04.
+         * Although the .keyIdentfier field is CONTEXT_SPECIFIC ([0] OPTIONAL),
+         * its tag with the content is the payload of on OCTET STRING primitive */
+        authority_key_id->keyIdentifier.tag = MBEDTLS_ASN1_OCTET_STRING;
+
+        *p += len;
+    }
```
```suggestion
    if (ret == 0) {
        authority_key_id->keyIdentifier.len = len;
        authority_key_id->keyIdentifier.p = *p;
        /* Setting tag of the keyIdentfier intentionally to 0x04.
         * Although the .keyIdentfier field is CONTEXT_SPECIFIC ([0] OPTIONAL),
         * its tag with the content is the payload of on OCTET STRING primitive */
        authority_key_id->keyIdentifier.tag = MBEDTLS_ASN1_OCTET_STRING;

        *p += len;
    } else if (ret != MBEDTLS_ERR_ASN1_UNEXPECTED_TAG) {
        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS, ret);
    }
```

DemiMarie reviewed (commented):
I can’t review the test data because I cannot read hex-encoded ASN.1 DER :laughing:.  The code (both library and test) looks pretty good, but I would prefer to see 100% branch coverage (exclusing OOM) for the parsing code.

DemiMarie reviewed (commented):
I can’t review the test data because I cannot read hex-encoded ASN.1 DER :laughing:.  The code (both library and test) looks pretty good, but I would prefer to see 100% branch coverage (exclusing OOM) for the parsing code.

DemiMarie reviewed (commented):
I can’t review the test data because I cannot read hex-encoded ASN.1 DER :laughing:.  The code (both library and test) looks pretty good, but I would prefer to see 100% branch coverage (exclusing OOM) for the parsing code.

mprse commented:
Coverage report can be found in fallow-up PR:
https://github.com/Mbed-TLS/mbedtls/pull/7411#issuecomment-1538455482

gilles-peskine-arm reviewed (approved):
I've reviewed the changes since 8194285cf1f2d5384f180e26b1924b299a56c09d which was [approved by Demi Marie](https://github.com/Mbed-TLS/mbedtls/pull/6866#pullrequestreview-1413660675) with reservations on the test data which will be improved in a follow-up. I approve the changes, again with reservations on the test data which will be improved in a follow-up.

gilles-peskine-arm commented:
The ABI-API-checking failures are changes in structure fields which are ok in a minor release. This is good to merge.
